feat: support for lambda event destinations (#1321)
awslabs/serverless-application-model: versions/2016-10-31.md

VersionDescription | `string` | A string that specifies the Description field which will be added on the new lambda version
ReservedConcurrentExecutions | `integer` | The maximum of concurrent executions you want to reserve for the function. For more information see [AWS Documentation on managing concurrency](https://docs.aws.amazon.com/lambda/latest/dg/concurrent-executions.html)
ProvisionedConcurrencyConfig | [ProvisionedConcurrencyConfig Object](#provisioned-concurrency-config-object) | Configure provisioned capacity for a number of concurrent executions on Lambda Alias property.

> EventInvokeConfig | [EventInvokeConfig object](#event-invoke-config-object) | Configure event destinations and retry behaviour on asynchronous invokes.

Reviewer: ```suggestion
EventInvokeConfig | [EventInvokeConfig object](#event-invoke-config-object) | Configure options for [asynchronous invocation](https://docs.aws.amazon.com/lambda/latest/dg/invocation-async.html) on the function.
```

---|:---:|---
ProvisionedConcurrentExecutions | `string` | Number of concurrent executions to be provisioned for the Lambda function. Required parameter.

#### Event Invoke Config object

The object describing event invoke config on a Lambda function. 

```yaml
  MyFunction:
    Type: 'AWS::Serverless::Function'
    Properties:
      EventInvokeConfig:
        MaximumEventAgeInSeconds: Integer (Min: 60, Max: 21600)
        MaximumRetryAttempts: Integer (Min: 0, Max: 2)
        DestinationConfig:
          OnSuccess:
            Type: [SQS | SNS | EventBridge | Function]
            Destination: ARN of [SQS | SNS | EventBridge | Function]
          OnFailure: 
            Type: [SQS | SNS | EventBridge | Function]
            Destination: ARN of [SQS | SNS | EventBridge | Function]
```

##### Properties
Property Name | Type | Description
---|:---:|---
MaximumEventAgeInSeconds | `integer` | The maximum age of a request that Lambda sends to a function for processing. Optional parameter.
MaximumRetryAttempts | `integer` | The maximum number of times to retry when the function returns an error. Optional parameter.

> DestinationConfig | [Desitnation Config Object](#event-invoke-destination-config-object) | A destination for events after they have been sent to a function for processing. Optional parameter.

Reviewer: ```suggestion
DestinationConfig | [Destination Config Object](#event-invoke-destination-config-object) | A destination for events after they have been sent to a function for processing. Optional parameter.
```

#### Event Invoke Destination Config object
The object describing destination config for Event Invoke Config.

##### Properties
Property Name | Type | Description
---|:---:|---

> OnSuccess | `object` | Takes Type of resource and Destination for events that succeeded processing.

Reviewer: Why are we just saying `object` here instead of defining a type? Since SAM has special behavior here (auto-creation of SNS/SQS), we should definitely document that as well as our support of `Fn::If` and the aliases they can use to Ref the auto-created queue/topic.

OnFailure | `object` | Takes Type of resource and Destination for events that failed processing.

##### Properties
Property Name | Type | Description
---|:---:|---

> ProvisionedConcurrentExecutions | `string` | Number of concurrent executions to be provisioned for the Lambda function. Required parameter.

Reviewer: This looks like a copy/paste error. Remove?
